src/proxy/http/canonicalize.rs

```diff
@@ -179,33 +188,62 @@ impl Future for Task {
 
     fn poll(&mut self) -> Poll<(), ()> {
         loop {
+            // If the receiver has been dropped, stop watching for updates.
+            match self.rx_stop.poll() {
+                Ok(Async::NotReady) => {}
+                _ => {
+                    trace!("task complete; name={:?}", self.original);
+                    return Ok(Async::Ready(()));
+                }
+            }
+
             self.state = match self.state {
                 State::Init => {
+                    trace!("task init; name={:?}", self.original);
                     let f = self.resolver.refine(self.original.name());
                     State::Pending(Timeout::new(f, self.timeout))
                 }
                 State::Pending(ref mut fut) => {
+                    // Only poll the resolution for updates when the receiver is
+                    // ready to receive an update.
+                    match self.tx.poll_ready() {
+                        Ok(Async::Ready(())) => {}
+                        Ok(Async::NotReady) => {
+                            trace!("task awaiting capacity; name={:?}", self.original);
+                            return Ok(Async::NotReady);
+                        }
+                        Err(_) => {
+                            trace!("task complete; name={:?}", self.original);
+                            return Ok(Async::Ready(()));
+                        }
+                    };
+
                     match fut.poll() {
                         Ok(Async::NotReady) => {
                             return Ok(Async::NotReady);
                         }
                         Ok(Async::Ready(refine)) => {
+                            trace!(
+                                "task update; name={:?} refined={:?}",
+                                self.original,
+                                refine.name
+                            );
                             // If the resolved name is a new name, bind a
                             // service with it and set a delay that will notify
                             // when the resolver should be consulted again.
                             let resolved = NameAddr::new(refine.name, self.original.port());
                             if self.resolved.get() != Some(&resolved) {
-                                let err = self.tx.try_send(resolved.clone()).err();
-                                if err.map(|e| e.is_disconnected()).unwrap_or(false) {
-                                    return Ok(().into());
-                                }
-
+                                self.tx
+                                    .try_send(resolved.clone())
+                                    .expect("tx failed despite being ready");
```

Comment:
Just to check my understanding: it is now impossible to get here if the receiver has been dropped because of the guard at the top of this function?

Reply:
Or doubly impossible because of the the `poll_ready` as well?

Reply:
The poll_ready is the important part of it. It guarantees there's a slot to send into.